Vanilla_CMFS_20NG.ipynb

```diff
@@ -0,0 +1,394 @@
+     ]
+    }
+   ],
+   "source": [
+    "print \"No. of documents: %d\\nNo. of categories: %d\" % (documents, categories)\n",
+    "print \"matrix.shape: {0}\".format(term_document_mat.shape)\n",
+    "print newsgroups_train.target[10]\n",
+    "print type(term_document_mat)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create Term-category feature-appearance matrix"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "term_category_mat = np.zeros((terms, categories))\n",
+    "for doc in range(documents):\n",
+    "    cat = newsgroups_train.target[doc]\n",
+    "    for row in term_document_mat.getcol(doc).nonzero()[0]:\n",
+    "        term_category_mat[row][cat] += 1"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Term-category matrix shape: (101323, 20)\n"
+     ]
+    }
+   ],
+   "source": [
+    "print \"Term-category matrix shape: {0}\".format(term_category_mat.shape)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Perform CMFS on term-category matrix"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[[  2.92719610e-08   7.28597222e-06   2.78309396e-06 ...,   5.88282467e-07\n",
+      "    2.72656836e-06   1.18472355e-07]\n",
+      " [  1.33768710e-06   2.15000369e-06   3.43217261e-07 ...,   5.26920017e-05\n",
+      "    1.01714536e-05   1.35350464e-06]\n",
+      " [  2.85130583e-07   2.77229556e-07   2.24044601e-07 ...,   9.16849120e-07\n",
+      "    2.65587955e-07   2.88502123e-07]\n",
+      " ..., \n",
+      " [  3.66596463e-07   3.56438001e-07   2.88057344e-07 ...,   2.94701503e-07\n",
+      "    3.41470227e-07   3.70931301e-07]\n",
+      " [  3.66596463e-07   3.56438001e-07   2.88057344e-07 ...,   2.94701503e-07\n",
+      "    3.41470227e-07   3.70931301e-07]\n",
+      " [  3.66596463e-07   3.56438001e-07   2.88057344e-07 ...,   2.94701503e-07\n",
+      "    3.41470227e-07   3.70931301e-07]]\n"
+     ]
+    }
+   ],
+   "source": [
+    "term_freq_per_cat = np.cumsum(term_category_mat, axis=0)[-1, :]\n",
+    "for term in range(terms):\n",
+    "    # Frequency of the term across all categories\n",
+    "    # Standard CMFS\n",
+    "    total_term_freq = sum(term_category_mat[term, :])\n",
+    "    for cat in range(categories):\n",
+    "        numerator = float((term_category_mat[term][cat] + 1) ** 2)\n",
+    "        denominator = (total_term_freq + categories) * (term_freq_per_cat[cat] + terms)\n",
+    "        term_category_mat[term][cat] = numerator / denominator\n",
```

Comment:
I didn't like this the first time I wrote this actually. We shouldn't be modifying the original term-category matrix. We would be needing another matrix so that the original matrix can be used for both, CMFS and ICMFS. After you obtain two different "modified" term-cat matrices from CMFS and ICMFS, it should be easy to go further.